### DISCUSSION TITLE
p-pickList

### DISCUSSION THREAD

**Discussion by esoftware-github** [primefaces/community#262] [PrimeNG] [answered]
I am using the component
```
<p-pickList
        [source]="roleList"
        [target]="assignedRolesList"
        sourceHeader="Available"
        targetHeader="Selected"
        [dragdrop]="true"
        [responsive]="true"
        [sourceStyle]="{ height: '20rem' }"
        [targetStyle]="{ height: '20rem' }"
        breakpoint="1400px"
        [showSourceControls]="false"
        [showTargetControls]="false"
      >
        <ng-template let-role pTemplate="item">
          {{ role.RoleName }}
        </ng-template>
      </p-pickList>
```

![image](https://github.com/primefaces/community/assets/70674019/e313c69c-178a-43e9-9a67-5f8a048b73e8)

When the page loads everything is fine

once the data is saved in the database and retrieved for editing, we populate the [target]="assignedRolesList" variable but the [source]="roleList" is not updated

```
const entity = await firstValueFrom(this.profileService.onGetById(parseInt(param)));
      // Agregar cada elemento de ProfileItems como un FormGroup al FormArray RoleList
      entity.ProfileItems.forEach((item: any) => {
        this.ProfileItemList.push(this.fb.group({
          ProfileItemID: item.ProfileItemID,
          ProfileID: item.ProfileID,
          ManageRoleID: item.ManageRoleID
        }));
      });


      this.assignedRolesList = entity.AssignedRolesList;

```
![image](https://github.com/primefaces/community/assets/70674019/4de42f6f-b1f5-48be-9eac-08094b7fb021)

thank you so much

**Comment by cetincakiroglu** [comment]
Hi,

Could you please create a stackblitz example so we can identify the issue precisely? It's related to change detection but we need an example to try possible workarounds without diving into the source code.

**Comment by cetincakiroglu** [comment]
Maybe this would help, created an example with to simulate request-response https://stackblitz.com/edit/fnkj9d?file=src%2Fapp%2Fdemo%2Fpicklist-filter-demo.ts,src%2Fapp%2Fdemo%2Fpicklist-filter-demo.html

**Reply by esoftware-github** [reply]
In the example, as you will see, the data has already been assigned to the target but the source remains the same, that is, the first 2 items in the source are already assigned to the target but they continue to be displayed in the source.


https://stackblitz.com/edit/fnkj9d-n3qsvk?file=src%2Fapp%2Fdemo%2Fpicklist-filter-demo.ts

[](https://stackblitz.com/edit/fnkj9d-n3qsvk?file=src%2Fapp%2Fdemo%2Fpicklist-filter-demo.ts)

**Reply by cetincakiroglu** [reply, marked as answer]
Hi,

The problem is change detection, you're copying the array and you need to modify it to trigger change detection or you can manually trigger change detection. See the example below

https://stackblitz.com/edit/fnkj9d-xbdeu1?file=src%2Fapp%2Fdemo%2Fpicklist-filter-demo.ts

**Comment by esoftware-github** [comment]
It is understood, in the case that in the loadData, there is the initial data with which the source will always be loaded and the idea is that when bringing the data from the database and assigning it to the target, the source must update with the differences with the target

![image](https://github.com/primefaces/community/assets/70674019/101a7fb4-725f-4917-a998-8790335d41e8)

[](https://stackblitz.com/edit/fnkj9d-7tqepj?file=src%2Fapp%2Fdemo%2Fpicklist-filter-demo.ts)

The solution I gave is this

`this.roleList = this.roleList.filter(item1 => !this.assignedRolesList.some(item2 => item2.ManageRoleID === item1.ManageRoleID));`


roleList  = source
assignedRolesList = from database

**Reply by cetincakiroglu** [reply]
Filter also creates a new array, the problem in your issue is you are not changing an array or creating a new one, you're copying the array that's why it doesn't detect the changes because there isn't a manipulation occurs. Filter or spread operator, you can use a lot of things to trigger change detection.
